File: pwiz/data/identdata/References.cpp
//
// $Id$
//
//
// Original author: Robert Burke <robert.burke@proteowizard.org>
//
// Copyright 2009 Spielberg Family Center for Applied Proteomics
//   University of Southern California, Los Angeles, California  90033
//
// Licensed under the Apache License, Version 2.0 (the "License"); 
// you may not use this file except in compliance with the License. 
// You may obtain a copy of the License at 
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software 
// distributed under the License is distributed on an "AS IS" BASIS, 
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
// See the License for the specific language governing permissions and 
// limitations under the License.
//

#define PWIZ_SOURCE

#include "References.hpp"
#include "TextWriter.hpp"
#include "pwiz/utility/misc/Std.hpp"

namespace pwiz {
namespace identdata {
namespace References {


template <typename object_type>
struct HasID
{
    const string& id_;
    HasID(const string& id) : id_(id) {}

    bool operator()(const shared_ptr<object_type>& objectPtr)
    {
        return objectPtr.get() && objectPtr->id == id_;
    }
};


template <typename object_type>
void resolve(shared_ptr<object_type>& reference, 
             const vector< shared_ptr<object_type> >& referentList)
{
    if (!reference.get() || reference->id.empty())
        return; 

    typename vector< shared_ptr<object_type> >::const_iterator it =
        find_if(referentList.begin(), referentList.end(), HasID<object_type>(reference->id));

    if (it == referentList.end())
    {
        ostringstream oss;
        oss << "[References::resolve()] Failed to resolve reference.\n"
            << "  object type: " << typeid(object_type).name() << endl
            << "  reference id: " << reference->id << endl
            << "  referent list: " << referentList.size() << endl;
        for (typename vector< shared_ptr<object_type> >::const_iterator it = referentList.begin();
            it != referentList.end(); ++it)
            oss << "    " << (*it)->id << endl;
        throw runtime_error(oss.str().c_str());
    }

    reference = *it;
}


template <typename object_type>
void resolve(vector < shared_ptr<object_type> >& references,
             const vector< shared_ptr<object_type> >& referentList)
{
    for (typename vector< shared_ptr<object_type> >::iterator it=references.begin();
         it!=references.end(); ++it)
        resolve(*it, referentList);
}

PWIZ_API_DECL void resolve(ContactRole& cr, IdentData& mzid)
{
    resolve(cr.contactPtr, mzid.auditCollection);
}


PWIZ_API_DECL void resolve(AnalysisSoftwarePtr& asp, IdentData& mzid)
{
    if (asp->contactRolePtr.get() && !asp->contactRolePtr->empty())
        resolve(*asp->contactRolePtr, mzid);
}


void resolve(Provider& provider, IdentData& mzid)
{   
    if (mzid.provider.contactRolePtr.get())
        resolve(*mzid.provider.contactRolePtr, mzid);
    if (mzid.provider.analysisSoftwarePtr.get())
        resolve(mzid.provider.analysisSoftwarePtr, mzid);
}


PWIZ_API_DECL void resolve(AnalysisSampleCollection& asc, IdentData& mzid)
{
    BOOST_FOREACH(SamplePtr& s, asc.samples)
    {
        BOOST_FOREACH(ContactRolePtr& cr, s->contactRole)
            resolve(*cr, mzid);
        BOOST_FOREACH(SamplePtr& ss, s->subSamples)
            if (ss.get() && !ss->empty())
                resolve(ss, asc.samples);
    }
}


PWIZ_API_DECL void resolve(OrganizationPtr& reference, vector<ContactPtr>& referentList)
{
    if (!reference.get() || reference->id.empty())
        return; 

    vector<ContactPtr>::iterator it = 
        find_if(referentList.begin(), referentList.end(), HasID<Contact>(reference->id));

    if (it == referentList.end())
    {
        ostringstream oss;
        oss << "[References::resolve()] Failed to resolve reference.\n"
            << "  object type: OrganizationPtr" << endl
            << "  reference id: " << reference->id << endl
            << "  referent list: " << referentList.size() << endl;
        for (vector<ContactPtr>::const_iterator it=referentList.begin(); it!=referentList.end(); ++it)
            oss << "    " << (*it)->id << endl;
        throw runtime_error(oss.str().c_str());
    }

    reference = boost::static_pointer_cast<Organization>(*it);
}


PWIZ_API_DECL void resolve(vector<ContactPtr>& vcp, IdentData& mzid)
{
    BOOST_FOREACH(ContactPtr& c, vcp)
    {
        if (dynamic_cast<Organization*>(c.get()))
            resolve(static_cast<Organization*>(c.get())->parent, mzid.auditCollection);
        else if (dynamic_cast<Person*>(c.get()))
            BOOST_FOREACH(OrganizationPtr& org, static_cast<Person*>(c.get())->affiliations)
                if (org.get() && !org->empty())
                    resolve(org, vcp);
    }
}

PWIZ_API_DECL void resolve(MassTablePtr& mt, const vector<SpectrumIdentificationProtocolPtr>& spectrumIdProts)
{
    if (!mt.get() || mt->id.empty())
        return; 

    BOOST_FOREACH(const SpectrumIdentificationProtocolPtr& sip, spectrumIdProts)
    BOOST_FOREACH(const MassTablePtr& mt2, sip->massTable)
    {
        if (mt == mt2)
            return;
        else if (mt->id == mt2->id)
        {
            mt = mt2;
            return;
        }
    }

    ostringstream oss;
    oss << "[References::resolve()] Failed to resolve reference.\n"
        << "  object type: MassTable" << endl
        << "  reference id: " << mt->id << endl
        << "  referent list: " << spectrumIdProts.size() << endl;
    for (vector<SpectrumIdentificationProtocolPtr>::const_iterator it=spectrumIdProts.begin();
         it!=spectrumIdProts.end(); ++it)
        oss << "    " << (*it)->id << endl;
    throw runtime_error(oss.str().c_str());
}

PWIZ_API_DECL void resolve(PeptideEvidencePtr& pe, const IdentData& mzid)
{
    if (!pe.get())
        throw runtime_error("NULL value passed into resolve(PeptideEvidencePtr, IdentData&)");

    if (!pe->peptidePtr || pe->peptidePtr->peptideSequence.empty())
        resolve(pe->peptidePtr, mzid.sequenceCollection.peptides);

    if (!pe->dbSequencePtr || (pe->dbSequencePtr->seq.empty() && pe->dbSequencePtr->length == 0))
        resolve(pe->dbSequencePtr, mzid.sequenceCollection.dbSequences);

    // TODO construct a collection of TranslationTable's from all the
    // SpectrumIdentificationProtocolPtr's in AnalysisProtocolCollection.
    
    //if (pe->translationTablePtr.get())
    //    resolve(pe->translationTablePtr, mzid.);
    
}


//template <IdentData& mzid>
struct ResolvePE
{
    const IdentData* mzid_;
    ResolvePE(const IdentData* mzid) : mzid_(mzid) {}
    
    void operator()(PeptideEvidencePtr& pe)
    {
        return resolve(pe, (*mzid_));
    }
};

PWIZ_API_DECL void resolve(SpectrumIdentificationListPtr& sil, IdentData& mzid)
{
    BOOST_FOREACH(SpectrumIdentificationResultPtr& sir, sil->spectrumIdentificationResult)
    {
        if (sir->spectraDataPtr.get())
            resolve(sir->spectraDataPtr, mzid.dataCollection.inputs.spectraData);

        BOOST_FOREACH(SpectrumIdentificationItemPtr& sii, sir->spectrumIdentificationItem)
        {
            resolve(sii->massTablePtr, mzid.analysisProtocolCollection.spectrumIdentificationProtocol);
            resolve(sii->samplePtr, mzid.analysisSampleCollection.samples);

            BOOST_FOREACH(IonTypePtr& it, sii->fragmentation)
            BOOST_FOREACH(FragmentArrayPtr& fa, it->fragmentArray)
                resolve(fa->measurePtr, sil->fragmentationTable);

            if (!mzid.sequenceCollection.empty() &&
                sii->peptidePtr.get() &&
                sii->peptidePtr->peptideSequence.empty())
            {
                resolve(sii->peptidePtr, mzid.sequenceCollection.peptides);
            }

            ResolvePE rpe(&mzid);
            for_each(sii->peptideEvidencePtr.begin(),
                     sii->peptideEvidencePtr.end(),
                     rpe);
        }
    }
}

PWIZ_API_DECL void resolve(SequenceCollection& sc, IdentData& mzid)
{
    BOOST_FOREACH(DBSequencePtr& dbs, sc.dbSequences)
        resolve(dbs->searchDatabasePtr, mzid.dataCollection.inputs.searchDatabase);
    
    // look for unresolved PeptidePtr and DBSequencePtr in PeptideEvidences

    bool needPeptideResolved = false, needDbSeqResolved = false;
    for (const auto& pe : sc.peptideEvidence)
    {
        if (!pe) throw runtime_error("NULL PeptideEvidencePtr");
        if (!pe->dbSequencePtr) throw runtime_error("NULL dbSequencePtr for PeptideEvidence (no id to resolve)");
        if (!pe->peptidePtr) throw runtime_error("NULL peptidePtr for PeptideEvidence (no id to resolve)");

        if (pe->dbSequencePtr->seq.empty() && pe->dbSequencePtr->length == 0)
            needDbSeqResolved = true;
        if (pe->peptidePtr->peptideSequence.empty())
            needPeptideResolved = true;

        if (needDbSeqResolved && needPeptideResolved)
            break;
    }

    map<string, const DBSequencePtr*> dbSeqById;
    if (needDbSeqResolved)
        for (const auto& s : sc.dbSequences)
            if (s)
                dbSeqById[s->id] = &s;

    map<string, const PeptidePtr*> peptideById;
    if (needPeptideResolved)
        for (const auto& p : sc.peptides)
            if (p)
                peptideById[p->id] = &p;

    for (auto& pe : sc.peptideEvidence)
    {
        // null check still needed in case loop above broke early
        if (!pe) throw runtime_error("NULL PeptideEvidencePtr");
        if (!pe->dbSequencePtr) throw runtime_error("NULL dbSequencePtr for PeptideEvidence (no id to resolve)");
        if (!pe->peptidePtr) throw runtime_error("NULL peptidePtr for PeptideEvidence (no id to resolve)");

        if (needDbSeqResolved && pe->dbSequencePtr->seq.empty() && pe->dbSequencePtr->length == 0)
        {
            auto findItr = dbSeqById.find(pe->dbSequencePtr->id);
            if (findItr == dbSeqById.end())
                throw runtime_error("dBSequence_ref for PeptideEvidence " + pe->id + " does not resolve to a DBSequence element");
            pe->dbSequencePtr = *findItr->second;
        }

        if (needPeptideResolved && pe->peptidePtr->peptideSequence.empty())
        {
            auto findItr = peptideById.find(pe->peptidePtr->id);
            if (findItr == peptideById.end())
                throw runtime_error("peptide_ref for PeptideEvidence " + pe->id + " does not resolve to a Peptide element");
            pe->peptidePtr = *findItr->second;
        }
    }

    // Create a single Enzyme referent list
    vector<EnzymePtr> enzymePtrs;
    BOOST_FOREACH(const SpectrumIdentificationProtocolPtr& sip, mzid.analysisProtocolCollection.spectrumIdentificationProtocol)
        enzymePtrs.insert(enzymePtrs.end(), sip->enzymes.enzymes.begin(), sip->enzymes.enzymes.end());
}

PWIZ_API_DECL void resolve(SpectrumIdentification& si, IdentData& mzid)
{
    if (si.spectrumIdentificationProtocolPtr.get())
        resolve(si.spectrumIdentificationProtocolPtr,
                mzid.analysisProtocolCollection.spectrumIdentificationProtocol);
    
    if (si.spectrumIdentificationListPtr.get() &&
        !mzid.dataCollection.analysisData.spectrumIdentificationList.empty())
        resolve(si.spectrumIdentificationListPtr,
                mzid.dataCollection.analysisData.spectrumIdentificationList);

    resolve(si.inputSpectra, mzid.dataCollection.inputs.spectraData);
    resolve(si.searchDatabase, mzid.dataCollection.inputs.searchDatabase);
}


PWIZ_API_DECL void resolve(AnalysisCollection& ac, IdentData& mzid)
{
    for (vector<SpectrumIdentificationPtr>::iterator it=ac.spectrumIdentification.begin();
         it != ac.spectrumIdentification.end(); it++)
        resolve(**it, mzid);

    // TODO resolve proteinDetectionProtocolPtr & proteinDetectionListPtr;
    resolve(ac.proteinDetection.proteinDetectionProtocolPtr,
            mzid.analysisProtocolCollection.proteinDetectionProtocol);

    if (ac.proteinDetection.proteinDetectionListPtr.get() &&
        mzid.dataCollection.analysisData.proteinDetectionListPtr.get())
    {
        if (ac.proteinDetection.proteinDetectionListPtr->id ==
            mzid.dataCollection.analysisData.proteinDetectionListPtr->id)
        {
            ac.proteinDetection.proteinDetectionListPtr =
                mzid.dataCollection.analysisData.proteinDetectionListPtr;
        }
        else 
            throw runtime_error("[References::resolve] Unresolved ProteinDetectionList");
    }

    if (!mzid.dataCollection.analysisData.spectrumIdentificationList.empty())
        resolve(ac.proteinDetection.inputSpectrumIdentifications,
                mzid.dataCollection.analysisData.spectrumIdentificationList);
}


PWIZ_API_DECL void resolve(vector<SpectrumIdentificationProtocolPtr>& vsip, IdentData& mzid)
{
    for (vector<SpectrumIdentificationProtocolPtr>::iterator it=vsip.begin();
         it!=vsip.end(); it++)
    {
        if (it->get())
            resolve((*it)->analysisSoftwarePtr, mzid.analysisSoftwareList);
    }
}


PWIZ_API_DECL void resolve(vector<ProteinDetectionProtocolPtr>& vpdp, IdentData& mzid)
{
    for (vector<ProteinDetectionProtocolPtr>::iterator it=vpdp.begin();
         it!=vpdp.end(); it++)
    {
        if (it->get())
            resolve((*it)->analysisSoftwarePtr, mzid.analysisSoftwareList);
    }    
}


PWIZ_API_DECL void resolve(DataCollection& dc, IdentData& mzid)
{
    BOOST_FOREACH(SpectrumIdentificationListPtr& sil, dc.analysisData.spectrumIdentificationList)
        resolve(sil, mzid);

    // If there's no proteinDetectionListPtr, then we're done.
    if (!dc.analysisData.proteinDetectionListPtr.get())
        return;

    // If SequenceCollection wasn't populated, then we're done.
    if (mzid.sequenceCollection.empty())
        return;

    ProteinDetectionListPtr pdl=dc.analysisData.proteinDetectionListPtr;

    BOOST_FOREACH(ProteinAmbiguityGroupPtr& pag, pdl->proteinAmbiguityGroup)
    {
        BOOST_FOREACH(ProteinDetectionHypothesisPtr& pdh, pag->proteinDetectionHypothesis)
        {
            resolve(pdh->dbSequencePtr, mzid.sequenceCollection.dbSequences);

            BOOST_FOREACH(PeptideHypothesis& ph, pdh->peptideHypothesis)
            {
                if (ph.peptideEvidencePtr && ph.peptideEvidencePtr->peptidePtr)
                    continue;
                resolve(ph.peptideEvidencePtr, mzid.sequenceCollection.peptideEvidence);

                //BOOST_FOREACH(SpectrumIdentificationItemPtr& sii, ph.spectrumIdentificationItemPtr)
                //    resolve(sii, mzid.analysisCollection.proteinDetection.inputSpectrumIdentifications);
            }
        }
    }
}


PWIZ_API_DECL void resolve(IdentData& mzid)
{
    BOOST_FOREACH(AnalysisSoftwarePtr& as, mzid.analysisSoftwareList)
        if (as->contactRolePtr.get())
            resolve(*as->contactRolePtr, mzid);

    resolve(mzid.provider, mzid);
    resolve(mzid.auditCollection, mzid);
    resolve(mzid.analysisSampleCollection, mzid);
    
    resolve(mzid.sequenceCollection, mzid);
    resolve(mzid.analysisCollection, mzid);
    resolve(mzid.analysisProtocolCollection.spectrumIdentificationProtocol, mzid);
    resolve(mzid.analysisProtocolCollection.proteinDetectionProtocol, mzid);
    resolve(mzid.dataCollection, mzid);
}


} // namespace References 
} // namespace identdata
} // namespace pwiz 

File: pwiz-sharp/pwiz/src/IdentData/References.cs
namespace Pwiz.Data.IdentData;

/// <summary>
/// Resolves cross-references inside an <see cref="IdentData"/> tree. Port of
/// <c>pwiz::identdata::References</c>.
/// </summary>
/// <remarks>
/// <para>mzIdentML uses id-references between elements: e.g. <c>SpectrumIdentificationItem</c>
/// references its source <c>Peptide</c> by <c>peptide_ref</c>. After parsing, schema objects
/// hold *stub* references (just <see cref="Identifiable.Id"/> populated) for any forward refs.
/// This utility walks the tree once to build id→object lookup tables and then walks again to
/// replace each stub with the fully-populated object from the table.</para>
/// <para>Idempotent: running <c>Resolve</c> twice produces the same tree. Callers who
/// build an IdentData programmatically (and don't want to plumb refs by hand) can populate the
/// list-owners (Peptides, DBSequences, etc.) and assign id-only stubs to cross-reference
/// fields, then call <c>Resolve</c> to link everything up.</para>
/// </remarks>
public static class References
{
    /// <summary>Resolves all id-references inside <paramref name="ident"/> in place.</summary>
    public static void Resolve(IdentData ident)
    {
        ArgumentNullException.ThrowIfNull(ident);
        var tables = LookupTables.From(ident);
        ResolveInto(ident, tables);
    }

    /// <summary>Resolves a single <see cref="ContactRole"/>'s contact reference.</summary>
    public static void Resolve(ContactRole cr, IdentData ident)
    {
        ArgumentNullException.ThrowIfNull(cr);
        ArgumentNullException.ThrowIfNull(ident);
        var tables = LookupTables.From(ident);
        ResolveContactRole(cr, tables);
    }

    private sealed class LookupTables
    {
        public Dictionary<string, AnalysisSoftware> Software { get; } = new(StringComparer.Ordinal);
        public Dictionary<string, Contact> Contact { get; } = new(StringComparer.Ordinal);
        public Dictionary<string, Organization> Organization { get; } = new(StringComparer.Ordinal);
        public Dictionary<string, Sample> Sample { get; } = new(StringComparer.Ordinal);
        public Dictionary<string, DBSequence> DbSequence { get; } = new(StringComparer.Ordinal);
        public Dictionary<string, Peptide> Peptide { get; } = new(StringComparer.Ordinal);
        public Dictionary<string, PeptideEvidence> PeptideEvidence { get; } = new(StringComparer.Ordinal);
        public Dictionary<string, SearchDatabase> SearchDb { get; } = new(StringComparer.Ordinal);
        public Dictionary<string, SpectraData> SpectraData { get; } = new(StringComparer.Ordinal);
        public Dictionary<string, SpectrumIdentificationList> SpectrumIdList { get; } = new(StringComparer.Ordinal);
        public Dictionary<string, SpectrumIdentificationProtocol> SpectrumIdProtocol { get; } = new(StringComparer.Ordinal);
        public Dictionary<string, ProteinDetectionProtocol> ProteinDetProtocol { get; } = new(StringComparer.Ordinal);
        public Dictionary<string, ProteinDetectionList> ProteinDetList { get; } = new(StringComparer.Ordinal);
        public Dictionary<string, Measure> Measure { get; } = new(StringComparer.Ordinal);

        public static LookupTables From(IdentData ident)
        {
            var t = new LookupTables();
            foreach (var sw in ident.AnalysisSoftwareList) Add(t.Software, sw.Id, sw);
            foreach (var c in ident.AuditCollection)
            {
                Add(t.Contact, c.Id, c);
                if (c is Organization o) Add(t.Organization, o.Id, o);
            }
            foreach (var s in ident.AnalysisSampleCollection.Samples) Add(t.Sample, s.Id, s);
            foreach (var dbs in ident.SequenceCollection.DBSequences) Add(t.DbSequence, dbs.Id, dbs);
            foreach (var p in ident.SequenceCollection.Peptides) Add(t.Peptide, p.Id, p);
            foreach (var pe in ident.SequenceCollection.PeptideEvidence) Add(t.PeptideEvidence, pe.Id, pe);
            foreach (var sip in ident.AnalysisProtocolCollection.SpectrumIdentificationProtocol) Add(t.SpectrumIdProtocol, sip.Id, sip);
            foreach (var pdp in ident.AnalysisProtocolCollection.ProteinDetectionProtocol) Add(t.ProteinDetProtocol, pdp.Id, pdp);
            foreach (var sd in ident.DataCollection.Inputs.SpectraData) Add(t.SpectraData, sd.Id, sd);
            foreach (var db in ident.DataCollection.Inputs.SearchDatabase) Add(t.SearchDb, db.Id, db);
            foreach (var sil in ident.DataCollection.AnalysisData.SpectrumIdentificationList)
            {
                Add(t.SpectrumIdList, sil.Id, sil);
                foreach (var m in sil.FragmentationTable) Add(t.Measure, m.Id, m);
            }
            if (ident.DataCollection.AnalysisData.ProteinDetectionListPtr is { } pdl)
                Add(t.ProteinDetList, pdl.Id, pdl);
            return t;
        }

        private static void Add<T>(Dictionary<string, T> map, string id, T value)
        {
            if (!string.IsNullOrEmpty(id)) map[id] = value;
        }
    }

    private static void ResolveInto(IdentData ident, LookupTables t)
    {
        foreach (var sw in ident.AnalysisSoftwareList)
            ResolveContactRole(sw.ContactRolePtr, t);

        ResolveContactRole(ident.Provider.ContactRolePtr, t);
        ident.Provider.AnalysisSoftwarePtr = ResolveRef(ident.Provider.AnalysisSoftwarePtr, t.Software);

        foreach (var c in ident.AuditCollection)
        {
            switch (c)
            {
                case Person p:
                    for (int i = 0; i < p.Affiliations.Count; i++)
                        p.Affiliations[i] = ResolveRefIpc(p.Affiliations[i], t.Organization)!;
                    break;
                case Organization o:
                    o.Parent = ResolveRefIpc(o.Parent, t.Organization);
                    break;
            }
        }

        foreach (var s in ident.AnalysisSampleCollection.Samples)
        {
            foreach (var cr in s.ContactRole) ResolveContactRole(cr, t);
            for (int i = 0; i < s.SubSamples.Count; i++)
                s.SubSamples[i] = ResolveRefIpc(s.SubSamples[i], t.Sample)!;
        }

        foreach (var dbs in ident.SequenceCollection.DBSequences)
            dbs.SearchDatabasePtr = ResolveRefIpc(dbs.SearchDatabasePtr, t.SearchDb);

        foreach (var pe in ident.SequenceCollection.PeptideEvidence)
        {
            pe.PeptidePtr = ResolveRefIpc(pe.PeptidePtr, t.Peptide);
            pe.DBSequencePtr = ResolveRefIpc(pe.DBSequencePtr, t.DbSequence);
        }

        foreach (var sip in ident.AnalysisProtocolCollection.SpectrumIdentificationProtocol)
            sip.AnalysisSoftwarePtr = ResolveRef(sip.AnalysisSoftwarePtr, t.Software);
        foreach (var pdp in ident.AnalysisProtocolCollection.ProteinDetectionProtocol)
            pdp.AnalysisSoftwarePtr = ResolveRef(pdp.AnalysisSoftwarePtr, t.Software);

        foreach (var si in ident.AnalysisCollection.SpectrumIdentification)
        {
            si.SpectrumIdentificationProtocolPtr = ResolveRef(si.SpectrumIdentificationProtocolPtr, t.SpectrumIdProtocol);
            si.SpectrumIdentificationListPtr = ResolveRefIpc(si.SpectrumIdentificationListPtr, t.SpectrumIdList);
            for (int i = 0; i < si.InputSpectra.Count; i++)
                si.InputSpectra[i] = ResolveRef(si.InputSpectra[i], t.SpectraData)!;
            for (int i = 0; i < si.SearchDatabase.Count; i++)
                si.SearchDatabase[i] = ResolveRefIpc(si.SearchDatabase[i], t.SearchDb)!;
        }

        var pd = ident.AnalysisCollection.ProteinDetection;
        pd.ProteinDetectionProtocolPtr = ResolveRef(pd.ProteinDetectionProtocolPtr, t.ProteinDetProtocol);
        pd.ProteinDetectionListPtr = ResolveRefIpc(pd.ProteinDetectionListPtr, t.ProteinDetList);
        for (int i = 0; i < pd.InputSpectrumIdentifications.Count; i++)
            pd.InputSpectrumIdentifications[i] = ResolveRefIpc(pd.InputSpectrumIdentifications[i], t.SpectrumIdList)!;

        foreach (var sil in ident.DataCollection.AnalysisData.SpectrumIdentificationList)
        {
            foreach (var sir in sil.SpectrumIdentificationResult)
            {
                sir.SpectraDataPtr = ResolveRef(sir.SpectraDataPtr, t.SpectraData);
                foreach (var sii in sir.SpectrumIdentificationItem)
                {
                    sii.PeptidePtr = ResolveRefIpc(sii.PeptidePtr, t.Peptide);
                    for (int i = 0; i < sii.PeptideEvidencePtr.Count; i++)
                        sii.PeptideEvidencePtr[i] = ResolveRefIpc(sii.PeptideEvidencePtr[i], t.PeptideEvidence)!;
                    foreach (var ion in sii.Fragmentation)
                        foreach (var fa in ion.FragmentArray)
                            fa.MeasurePtr = ResolveRefIpc(fa.MeasurePtr, t.Measure);
                }
            }
        }

        if (ident.DataCollection.AnalysisData.ProteinDetectionListPtr is { } proteinList)
        {
            foreach (var pag in proteinList.ProteinAmbiguityGroup)
            {
                foreach (var pdh in pag.ProteinDetectionHypothesis)
                {
                    pdh.DBSequencePtr = ResolveRefIpc(pdh.DBSequencePtr, t.DbSequence);
                    foreach (var ph in pdh.PeptideHypothesis)
                        ph.PeptideEvidencePtr = ResolveRefIpc(ph.PeptideEvidencePtr, t.PeptideEvidence);
                }
            }
        }
    }

    private static void ResolveContactRole(ContactRole? cr, LookupTables t)
    {
        if (cr is null) return;
        cr.ContactPtr = ResolveRefIpc(cr.ContactPtr, t.Contact);
    }

    /// <summary>Replaces a stub reference (one with a non-empty Id but otherwise empty) with the
    /// populated entry from <paramref name="map"/>. Returns the populated entry when found,
    /// otherwise returns the original (which may be a stub or null).</summary>
    private static T? ResolveRef<T>(T? stub, Dictionary<string, T> map) where T : Identifiable
    {
        if (stub is null) return null;
        if (string.IsNullOrEmpty(stub.Id)) return stub;
        return map.TryGetValue(stub.Id, out var hit) ? hit : stub;
    }

    /// <summary>Same as <see cref="ResolveRef{T}(T, Dictionary{string, T})"/> but for
    /// <see cref="IdentifiableParamContainer"/> descendants — separate overload because the two
    /// base classes don't share a common ancestor with an Id property in pwiz-sharp's port.</summary>
    private static T? ResolveRefIpc<T>(T? stub, Dictionary<string, T> map) where T : IdentifiableParamContainer
    {
        if (stub is null) return null;
        if (string.IsNullOrEmpty(stub.Id)) return stub;
        return map.TryGetValue(stub.Id, out var hit) ? hit : stub;
    }
}
